src/decaylanguage/utils/utilities.py

```diff
@@ -45,12 +45,13 @@ def split(x):  # type: ignore [no-untyped-def]
     return out
 
 
-def filter_lines(matcher, inp):  # type: ignore [no-untyped-def]
+def filter_lines(
+    matcher: Pattern[str], inp: List[str]
+) -> Tuple[List[Dict[str, Union[str, Any]]], List[str]]:
     """
     Filter out lines into new variable if they match a regular expression
     """
-    output = [
-        matcher.match(ln).groupdict() for ln in inp if matcher.match(ln) is not None
-    ]
+    matches = (matcher.match(ln) for ln in inp)
+    output = [match.groupdict() for match in matches if match is not None]
```

Comment:
Reworked this to make mypy happy (and it's technically a line shorter now ;) ) - mypy doesn't know that `matcher.match(ln)` is the same when you run it twice. Without Python 3.8 walrus operator, the other way to fix it was to pull it out into a generator expression to avoid the double call, which I did.

Reply:
Nice.

Reply:
FYI, the walrus version would have been:

```python
output = [
    match.groupdict() for ln in inp if (match := matcher.match(ln)) is not None
]
```

As is often the case, I think the two line version is more readable than the walrus.

Reply:
Indeed not too bad but the 2-line version might actually be quicker to read.